processing/src/main/java/io/druid/query/dimension/LookupDimensionSpec.java

```diff
@@ -0,0 +1,251 @@
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.metamx.common.StringUtils;
+import io.druid.query.extraction.DimExtractionFn;
+import io.druid.query.extraction.ExtractionFn;
+import io.druid.query.extraction.LookupExtractor;
+import io.druid.query.extraction.LookupReferencesManager;
+import io.druid.query.filter.DimFilterCacheHelper;
+import io.druid.segment.DimensionSelector;
+
+import javax.annotation.Nullable;
+import java.nio.ByteBuffer;
+
+public class LookupDimensionSpec implements DimensionSpec
+{
+  private static final byte CACHE_TYPE_ID = 0x4;
+
+  @JsonProperty
+  private final String dimension;
+
+  @JsonProperty
+  private final String outputName;
+
+  @JsonProperty
+  private final LookupExtractor lookup;
+
+  @JsonProperty
+  private final boolean retainMissingValues;
+
+  @JsonProperty
+  private final String replaceMissingWith;
+
+  @JsonProperty
+  private final String name;
+
+
+  private final LookupReferencesManager lookupReferencesManager;
+
+
+  @JsonCreator
+  public LookupDimensionSpec(
+      @JsonProperty("dimension") String dimension,
+      @JsonProperty("outputName") String outputName,
+      @JsonProperty("lookup") LookupExtractor lookup,
+      @JsonProperty("retainMissingValues") final boolean retainMissingValues,
+      @JsonProperty("replaceMissingWith") final String replaceMissingWith,
+      @JsonProperty("name") String name,
+      @JacksonInject LookupReferencesManager lookupReferencesManager
+  )
+  {
+    this.retainMissingValues = retainMissingValues;
+    this.replaceMissingWith = Strings.emptyToNull(replaceMissingWith);
+    Preconditions.checkArgument(
+        !(retainMissingValues && !Strings.isNullOrEmpty(replaceMissingWith)),
+        "Cannot specify a [replaceMissingValue] and set [retainMissingValue] to true"
+    );
+    this.dimension = Preconditions.checkNotNull(dimension, "dimension can not be Null");
+    this.outputName = Preconditions.checkNotNull(outputName, "outputName can not be Null");
+    this.lookupReferencesManager = lookupReferencesManager;
+    this.name = name;
+    this.lookup = lookup;
+    Preconditions.checkArgument(
+        Strings.isNullOrEmpty(name) ^ (lookup == null),
+        "name and lookup are mutually exclusive please provide either a name or a lookup"
+    );
+  }
+
+  @Override
+  @JsonProperty
+  public String getDimension()
+  {
+    return dimension;
+  }
+
+  @Override
+  @JsonProperty
+  public String getOutputName()
+  {
+    return outputName;
+  }
+
+  @JsonProperty
+  @Nullable
+  public LookupExtractor getLookup()
+  {
+    return lookup;
+  }
+
+  @JsonProperty
+  @Nullable
+  public String getName()
+  {
+    return name;
+  }
+
+  @Override
+  public ExtractionFn getExtractionFn()
+  {
+    final LookupExtractor lookupExtractor = Strings.isNullOrEmpty(name) ? this.lookup
+                                                                        : Preconditions.checkNotNull(
```

Comment:
I think this code would be more readable as a series of if statements instead of using the `Preconditions` stuff.